test/cpu_adaptiveavgpool.hpp

```diff
@@ -0,0 +1,311 @@
+        float sum     = 0;
+        for(size_t ih = h; ih < (h + kh); ++ih)
+        {
+            for(size_t iw = w; iw < (w + kw); ++iw)
+            {
+                sum += static_cast<float>(input[input_tv.get_tensor_view_idx({n, c, ih, iw})]);
+            }
+        }
+
+        output[output_tv.get_tensor_view_idx({n, c, oh, ow})] = static_cast<T>(sum / divider);
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_forward_3d(tensor<T> input,
+                                    tensor<T>& output,
+                                    size_t N,
+                                    size_t C,
+                                    size_t D,
+                                    size_t H,
+                                    size_t W,
+                                    size_t OD,
+                                    size_t OH,
+                                    size_t OW)
+{
+    auto dims  = input.desc.GetLengths();
+    auto numel = output.desc.GetElementSize();
+
+    auto input_tv  = miopen::get_inner_expanded_tv<5>(input.desc);
+    auto output_tv = miopen::get_inner_expanded_tv<5>(output.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
+    {
+        size_t ncodoh = gid / OW, ow = gid % OW;
+        size_t ncod = ncodoh / OH, oh = ncodoh % OH;
+        size_t nc = ncod / OD, od = ncod % OD;
+        size_t n = nc / C, c = nc % C;
+
+        if(n >= N)
+            return;
+
+        size_t d  = static_cast<size_t>(std::floor(static_cast<float>(od * D) / OD));
+        size_t kd = static_cast<size_t>(std::ceil(static_cast<float>((od + 1) * D) / OD)) - d;
+
+        size_t h  = static_cast<size_t>(std::floor(static_cast<float>(oh * H) / OH));
+        size_t kh = static_cast<size_t>(std::ceil(static_cast<float>((oh + 1) * H) / OH)) - h;
+
+        size_t w  = static_cast<size_t>(std::floor(static_cast<float>(ow * W) / OW));
+        size_t kw = static_cast<size_t>(std::ceil(static_cast<float>((ow + 1) * W) / OW)) - w;
+
+        float sum = 0;
+        for(size_t id = d; id < (d + kd); ++id)
+        {
+            for(size_t ih = h; ih < (h + kh); ++ih)
+            {
+                for(size_t iw = w; iw < (w + kw); ++iw)
+                {
+                    sum +=
+                        static_cast<float>(input[input_tv.get_tensor_view_idx({n, c, id, ih, iw})]);
+                }
+            }
+        }
+
+        output[output_tv.get_tensor_view_idx({n, c, od, oh, ow})] =
+            static_cast<T>(sum / (kd * kh * kw));
+    }
+}
+
+template <class T>
+void cpu_adaptiveavgpool_backward_1d(
+    tensor<T> output_grad, tensor<T>& input_grad, size_t N, size_t C, size_t H, size_t OH)
+{
+    auto dims  = input_grad.desc.GetLengths();
+    auto numel = input_grad.desc.GetElementSize();
+
+    auto output_grad_tv = miopen::get_inner_expanded_tv<3>(output_grad.desc);
+    auto input_grad_tv  = miopen::get_inner_expanded_tv<3>(input_grad.desc);
+
+    for(size_t gid = 0; gid < numel; gid++)
```

Comment:
You should use `par_ford` to perform parallel calculation, for better performance.